src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java

```diff
@@ -1285,68 +1096,64 @@ private Name emitTryFinally(int pos) {
-        mv.visitLabel(lTo);
+        cob.labelBinding(lFrom);
+        emitPushArgument(cob, invoker, 0); // load target
+        emitPushArguments(cob, args, 1); // load args (skip 0: method handle)
+        cob.invokevirtual(CD_MethodHandle, "invokeBasic", methodDesc(type.basicType()));
+        cob.labelBinding(lTo);
 
         // FINALLY_NORMAL:
         int index = extendLocalsMap(new Class<?>[]{ returnType });
         if (isNonVoid) {
-            emitStoreInsn(basicReturnType, index);
+            emitStoreInsn(cob, basicReturnType.basicTypeKind(), index);
         }
-        emitPushArgument(invoker, 1); // load cleanup
-        mv.visitInsn(Opcodes.ACONST_NULL);
+        emitPushArgument(cob, invoker, 1); // load cleanup
+        cob.loadConstant(null);
         if (isNonVoid) {
-            emitLoadInsn(basicReturnType, index);
+            emitLoadInsn(cob, basicReturnType.basicTypeKind(), index);
         }
-        emitPushArguments(args, 1); // load args (skip 0: method handle)
-        mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL, MH, "invokeBasic", cleanupDesc, false);
-        mv.visitJumpInsn(Opcodes.GOTO, lDone);
+        emitPushArguments(cob, args, 1); // load args (skip 0: method handle)
+        cob.invokevirtual(CD_MethodHandle, "invokeBasic", cleanupDesc);
+        cob.goto_w(lDone);
 
         // CATCH:
-        mv.visitLabel(lCatch);
-        mv.visitInsn(Opcodes.DUP);
+        cob.labelBinding(lCatch);
+        cob.dup();
 
         // FINALLY_EXCEPTIONAL:
-        emitPushArgument(invoker, 1); // load cleanup
-        mv.visitInsn(Opcodes.SWAP);
+        emitPushArgument(cob, invoker, 1); // load cleanup
+        cob.swap();
         if (isNonVoid) {
-            emitZero(BasicType.basicType(returnType)); // load default for result
+            emitZero(cob, BasicType.basicType(returnType)); // load default for result
         }
-        emitPushArguments(args, 1); // load args (skip 0: method handle)
-        mv.visitMethodInsn(Opcodes.INVOKEVIRTUAL, MH, "invokeBasic", cleanupDesc, false);
+        emitPushArguments(cob, args, 1); // load args (skip 0: method handle)
+        cob.invokevirtual(CD_MethodHandle, "invokeBasic", cleanupDesc);
         if (isNonVoid) {
-            emitPopInsn(basicReturnType);
+            emitPopInsn(cob, basicReturnType);
         }
-        mv.visitInsn(Opcodes.ATHROW);
+        cob.athrow();
 
         // DONE:
-        mv.visitLabel(lDone);
+        cob.labelBinding(lDone);
 
         return result;
     }
 
-    private void emitPopInsn(BasicType type) {
-        mv.visitInsn(popInsnOpcode(type));
-    }
-
-    private static int popInsnOpcode(BasicType type) {
-        return switch (type) {
-            case I_TYPE, F_TYPE, L_TYPE -> Opcodes.POP;
-            case J_TYPE, D_TYPE         -> Opcodes.POP2;
+    private void emitPopInsn(CodeBuilder cob, BasicType type) {
```

Comment:
We need a TypeKind-aware pop in CodeBuilder too :(

Reply:
Yes, it would be helpful.